### PR TITLE
Fix WinGet update loop for packages whose upgrade never sticks

### LINKED ISSUE #5158: [BUG] Update Loop (Unkown)
### Please confirm these before moving forward

- [x] I have searched for my issue and have not found a work-in-progress/duplicate/resolved issue.
- [x] I have tested that this issue has not been fixed in the latest [(beta or stable) release](https://github.com/Devolutions/UniGetUI/releases/).
- [x] I have checked the [FAQ](https://github.com/Devolutions/UniGetUI#frequently-asked-questions) section for solutions.
- [x] This issue is about a bug (if it is not, please use the [correct template](https://github.com/Devolutions/UniGetUI/issues/new/choose)).

### UniGetUI Version

2026.2.5

### Windows version, edition, and architecture

Windows 11 IoT 26100.8875 x 64

### Describe your issue

For the past few days I have been having issues with the program, and I have been "updating" over and over the same programs over and over again. I also tried uninstalling and installing the apps, and they still show the same error. If you need more for me to show, do tell me. And I do apologize if this was an... easy fix.

### Steps to reproduce the issue

_No response_

### UniGetUI Log

```text
[20/7/2026 17:22:03] Found 8 available updates from Winget
[20/7/2026 17:22:03] Could not resolve installer hosts for ShiftUp.NIKKE version Unknown: Failed during source operation 'manifest' for 'winget': Version 'Unknown' was not found for package 'ShiftUp.NIKKE' in source 'winget'.
[20/7/2026 17:22:04] Could not resolve installer hosts for Smilegate.Stove version Unknown: Failed during source operation 'manifest' for 'winget': Version 'Unknown' was not found for package 'Smilegate.Stove' in source 'winget'.
[20/7/2026 17:22:04] Could not resolve installer hosts for c0re100.qBittorrent-Enhanced-Edition version 5.2.3: Failed during source operation 'manifest' for 'winget': Version '5.2.3' was not found for package 'c0re100.qBittorrent-Enhanced-Edition' in source 'winget'.
[20/7/2026 17:22:21] Unknown mimetype image/jpeg for icon https://cdn.brandfetch.io/idQEkAXA-z/w/400/h/400/theme/dark/icon.jpeg?c=1bxid64Mup7aczewSAYMX&t=1769771355360, aborting download
[20/7/2026 17:22:28] Loaded operation parameters for package id=Microsoft.VCRedist.2013.x64 on manager Winget and operation Update: update --id "Microsoft.VCRedist.2013.x64" --exact --source winget --silent --include-unknown --architecture x64
[20/7/2026 17:22:30] Loaded operation parameters for package id=Microsoft.VCRedist.2013.x86 on manager Winget and operation Update: update --id "Microsoft.VCRedist.2013.x86" --exact --source winget --silent --include-unknown --architecture x86
[20/7/2026 17:22:30] Loaded operation parameters for package id=Microsoft.DotNet.DesktopRuntime.8 on manager Winget and operation Update: update --id "Microsoft.DotNet.DesktopRuntime.8" --exact --source winget --silent --include-unknown
[20/7/2026 17:22:30] Loaded operation parameters for package id=Microsoft.DotNet.DesktopRuntime.9 on manager Winget and operation Update: update --id "Microsoft.DotNet.DesktopRuntime.9" --exact --source winget --silent --include-unknown
[20/7/2026 17:22:31] Loaded operation parameters for package id=ShiftUp.NIKKE on manager Winget and operation Update: update --id "ShiftUp.NIKKE" --exact --source winget --silent --include-unknown
[20/7/2026 17:22:31] Loaded operation parameters for package id=c0re100.qBittorrent-Enhanced-Edition on manager Winget and operation Update: update --id "c0re100.qBittorrent-Enhanced-Edition" --exact --source winget --silent --include-unknown
[20/7/2026 17:22:31] Loaded operation parameters for package id=Smilegate.Stove on manager Winget and operation Update: update --id "Smilegate.Stove" --exact --source winget --silent --include-unknown
[20/7/2026 17:22:31] Loaded operation parameters for package id=Tailscale.Tailscale on manager Winget and operation Update: update --id "Tailscale.Tailscale" --exact --source winget --silent --include-unknown
[20/7/2026 17:22:32] Version string v0.41.0-867-g4c220ffd9 appears to contain non-numeric characters within a numeric segment and will be treated as unknown
[20/7/2026 17:22:32] Version string v0.41.0-867-g4c220ffd9 appears to contain non-numeric characters within a numeric segment and will be treated as unknown
[20/7/2026 17:22:32] Found 136 installed packages from Winget
```

### Package Managers Logs

```text
Manager WinGet with version:
Pinget CLI Version: pinget 0.10.0
Using Pinget CLI helper (JSON parsing)

——————————————————————————————————————————

Logged subprocess-based task on manager Winget. Task type is ListSources
Subprocess executable: "C:\Users\Alan Zamarron\scoop\apps\unigetui\current\pinget.exe"
Command-line arguments: " source export --output json"
Process start time: 20/7/2026 17:21:56
Process end time:   20/7/2026 17:21:56

-- Process STDOUT
 ...

Return code: SUCCESS (0)

——————————————————————————————————————————
Logged subprocess-based task on manager Winget. Task type is RefreshIndexes
Subprocess executable: "C:\Users\Alan Zamarron\scoop\apps\unigetui\current\pinget.exe"
Command-line arguments: " source update "
Process start time: 20/7/2026 17:21:57
Process end time:   20/7/2026 17:21:58

-- Process STDOUT
 ...

Return code: SUCCESS (0)

——————————————————————————————————————————
Logged subprocess-based task on manager Winget. Task type is ListInstalledPackages
Subprocess executable: "C:\Users\Alan Zamarron\scoop\apps\unigetui\current\pinget.exe"
Command-line arguments: " list --output json"
Process start time: 20/7/2026 17:21:58
Process end time:   20/7/2026 17:22:00

-- Process STDOUT
 ...

Return code: SUCCESS (0)

——————————————————————————————————————————
Logged subprocess-based task on manager Winget. Task type is ListUpdates
Subprocess executable: "C:\Users\Alan Zamarron\scoop\apps\unigetui\current\pinget.exe"
Command-line arguments: " upgrade --include-unknown --output json"
Process start time: 20/7/2026 17:22:00
Process end time:   20/7/2026 17:22:03

-- Process STDOUT
 ...

Return code: SUCCESS (0)

——————————————————————————————————————————
Logged subprocess-based task on manager Winget. Task type is ListInstalledPackages
Subprocess executable: "C:\Users\Alan Zamarron\scoop\apps\unigetui\current\pinget.exe"
Command-line arguments: " list --output json"
Process start time: 20/7/2026 17:22:30
Process end time:   20/7/2026 17:22:32

-- Process STDOUT
 ...

Return code: SUCCESS (0)

——————————————————————————————————————————
Logged subprocess-based task on manager Winget. Task type is ListInstalledPackages
Subprocess executable: "C:\Users\Alan Zamarron\scoop\apps\unigetui\current\pinget.exe"
Command-line arguments: " list --output json"
Process start time: 20/7/2026 17:22:32
Process end time:   20/7/2026 17:22:34

-- Process STDOUT
 ...

Return code: SUCCESS (0)

——————————————————————————————————————————
Logged subprocess-based task on manager Winget. Task type is ListInstalledPackages
Subprocess executable: "C:\Users\Alan Zamarron\scoop\apps\unigetui\current\pinget.exe"
Command-line arguments: " list --output json"
Process start time: 20/7/2026 17:22:34
Process end time:   20/7/2026 17:22:37

-- Process STDOUT
 ...

Return code: SUCCESS (0)

——————————————————————————————————————————
Logged subprocess-based task on manager Winget. Task type is ListInstalledPackages
Subprocess executable: "C:\Users\Alan Zamarron\scoop\apps\unigetui\current\pinget.exe"
Command-line arguments: " list --output json"
Process start time: 20/7/2026 17:22:37
Process end time:   20/7/2026 17:22:39

-- Process STDOUT
 ...

Return code: SUCCESS (0)

——————————————————————————————————————————
Logged subprocess-based task on manager Winget. Task type is ListInstalledPackages
Subprocess executable: "C:\Users\Alan Zamarron\scoop\apps\unigetui\current\pinget.exe"
Command-line arguments: " list --output json"
Process start time: 20/7/2026 17:22:39
Process end time:   20/7/2026 17:22:41

-- Process STDOUT
 ...

Return code: SUCCESS (0)

——————————————————————————————————————————
Logged subprocess-based task on manager Winget. Task type is ListInstalledPackages
Subprocess executable: "C:\Users\Alan Zamarron\scoop\apps\unigetui\current\pinget.exe"
Command-line arguments: " list --output json"
Process start time: 20/7/2026 17:22:41
Process end time:   20/7/2026 17:22:43

-- Process STDOUT
 ...

Return code: SUCCESS (0)

——————————————————————————————————————————
Logged subprocess-based task on manager Winget. Task type is ListInstalledPackages
Subprocess executable: "C:\Users\Alan Zamarron\scoop\apps\unigetui\current\pinget.exe"
Command-line arguments: " list --output json"
Process start time: 20/7/2026 17:22:43
Process end time:   20/7/2026 17:22:45

-- Process STDOUT
 ...

Return code: SUCCESS (0)

——————————————————————————————————————————
Logged subprocess-based task on manager Winget. Task type is ListInstalledPackages
Subprocess executable: "C:\Users\Alan Zamarron\scoop\apps\unigetui\current\pinget.exe"
Command-line arguments: " list --output json"
Process start time: 20/7/2026 17:22:45
Process end time:   20/7/2026 17:22:47

-- Process STDOUT
 ...

Return code: SUCCESS (0)
```

### Relevant information

_No response_

### Screenshots and videos

Image

Image

### PR DESCRIPTION
**Pull request by GabrielDuf**
This pull request implements a robust solution to suppress "phantom" WinGet package updates that repeatedly appear as available but never actually install (issue #5158). The main changes introduce a mechanism to track and suppress updates that have been "successfully" upgraded multiple times without the installed version advancing, while ensuring legitimate updates are still shown. Additional improvements include falling back to the last upgraded version when WinGet reports an "Unknown" installed version, and comprehensive tests to cover these scenarios.

**Suppression of phantom WinGet updates and version restoration:**

* Added tracking of upgrade attempts in `WinGetPkgOperationHelper` using a new `WinGetUpgradeAttempts` setting. If an update is "successfully" applied a configurable number of times (default: 3) without the installed version changing, it is persistently suppressed until a newer version is available. 
* Modified the logic in `ConsumeAlreadyUpgradedSuppression` to use this loop breaker, and log when a package is suppressed due to being stuck.

**Version handling improvements:**

* Implemented `ResolveReportedInstalledVersion` to restore the last upgraded version when WinGet reports "Unknown", ensuring the package appears up-to-date if it was just upgraded. 
* Updated all WinGet manager helpers (`PingetCliHelper`, `WinGetCliHelper`, `NativeWinGetHelper`) to use the new version resolution logic and properly suppress stuck updates. 

**Settings and configuration:**

* Added `WinGetUpgradeAttempts` to the settings engine, including import/export support and key resolution. 

**Test coverage:**

* Added comprehensive tests for stuck upgrade loop detection, version restoration, and correct suppression behavior, including edge cases like reboot-required upgrades.

**Bug fixes and edge case handling:**

* Ensured that upgrades requiring a reboot do not count toward the stuck loop threshold, preventing false suppression.

These changes collectively address persistent phantom update notifications, improve user experience, and ensure the update status accurately reflects the real state of WinGet-managed packages.

### CODE REVIEW

**Comment by Kobi-Blade** [comment]
I reviewed the UniGet WinGet helper code and confirmed that UniGet caches both `WindowsPackageManagerFactory` and `PackageManager globally`:

This means the same `PackageManager` instance is reused across all operations?

Microsoft does not publish documentation explicitly forbidding this, but reusing the same instance can lead to this stale catalogs, installed versions, and phantom updates (had problems with this in past).

The changes in this PR improve how UniGet handles phantom updates, but they might not fix the underlying issue.

**Comment by GabrielDuf** [comment]
@Kobi-Blade,
Thanks for digging into this. You’re right that the COM path reuses a single PackageManager, and the update list uses a cached local-package snapshot. That’s worth reviewing separately.

However, it isn’t causing this issue. The reporter is using the bundled Pinget CLI (pinget 0.10.0), where UniGetUI does not create the COM helper. Each list or upgrade operation starts a fresh pinget.exe process with a new system view.

The loop happens because the detected installed version never matches the catalog version, or cannot be read:

VCRedist 2013: 12.0.30501 installed vs 12.0.40664 offered
qBittorrent Enhanced: 5.2.3 installed vs 5.2.3.10 offered
NIKKE/STOVE: installed version reported as Unknown

Each upgrade reports success without changing that state, so refreshing the COM instance or catalog would return the same result.

This PR targets updates that cannot be confirmed as applied. The COM instance-reuse concern is valid, but it is a separate issue.

**Comment by Kobi-Blade** [comment]
Thanks for the clear explanation.

That makes complete sense, I missed that this was targeting Pinget.

File: src/UniGetUI.PackageEngine.Managers.WinGet/ClientHelpers/PingetCliHelper.cs [hunk 1]
```diff
@@ -46,19 +46,29 @@ public IReadOnlyList<Package> GetAvailableUpdates_UnSafe()
...
+                    ? WinGetPkgOperationHelper.GetLastInstalledVersion(match.Id)
+                    : match.InstalledVersion,
                 match.AvailableVersion!,
                 GetSource(match),
                 Manager
             );
 
-            if (!WinGetPkgOperationHelper.ConsumeAlreadyUpgradedSuppression(package))
+            // Skip one-shot suppression for unknown versions so the restored mark isn't cleared.
+            if (versionUnknown || !WinGetPkgOperationHelper.ConsumeAlreadyUpgradedSuppression(package))
```
[the remark this answers is not in the document]
Comment (GabrielDuf) [reply src/UniGetUI.PackageEngine.Managers.WinGet/ClientHelpers/PingetCliHelper.cs:71]: false positive : This one doesn't surface a visible phantom. An already-upgraded unknown package is returned with VersionString == NewVersionString (the restored version equals the available one),   
  and UpgradablePackagesLoader.IsPackageValid drops exactly those before anything is shown (if (package.VersionString == package.NewVersionString) return false;). The manager helpers  
  intentionally return candidates and let the loader filter — matching the existing COM/CLI behavior — so no change here.

File: src/UniGetUI.PackageEngine.Managers.WinGet/ClientHelpers/WinGetCliHelper.cs [hunk 2]
```diff
@@ -155,8 +155,14 @@ public IReadOnlyList<Package> GetAvailableUpdates_UnSafe()
...
 
+                // Restore the version we last upgraded to when WinGet reports it as unknown (#5158).
+                bool versionUnknown = WinGetPkgOperationHelper.IsUnknownVersion(version);
+                if (versionUnknown)
+                    version = WinGetPkgOperationHelper.GetLastInstalledVersion(id);
+
                 var package = new Package(name, id, version, newVersion, source, Manager);
-                if (!WinGetPkgOperationHelper.ConsumeAlreadyUpgradedSuppression(package))
+                // Skip one-shot suppression for unknown versions so the restored mark isn't cleared.
+                if (versionUnknown || !WinGetPkgOperationHelper.ConsumeAlreadyUpgradedSuppression(package))
```
[the remark this answers is not in the document]
Comment (GabrielDuf) [reply src/UniGetUI.PackageEngine.Managers.WinGet/ClientHelpers/WinGetCliHelper.cs:165]: Same as the Pinget path: the restored version equals the available version, and UpgradablePackagesLoader.IsPackageValid filters VersionString == NewVersionString before display. IsUpgradable being true doesn't matter because the loader gate, not that flag, decides visibility. Left as-is for consistency with how all three helpers feed the loader.

File: src/UniGetUI.PackageEngine.Managers.WinGet/ClientHelpers/NativeWinGetHelper.cs [hunk 3]
```diff
@@ -349,9 +349,10 @@ public IReadOnlyList<Package> GetAvailableUpdates_UnSafe()
                     nativePackage.DefaultInstallVersion.PackageCatalog.Info.Name
                 );
 
-                string version = nativePackage.InstalledVersion.Version;
-                if (version == "Unknown")
-                    version = WinGetPkgOperationHelper.GetLastInstalledVersion(nativePackage.Id);
+                string version = WinGetPkgOperationHelper.ResolveReportedInstalledVersion(
+                    nativePackage.Id,
+                    nativePackage.InstalledVersion.Version
+                );
```
[the remark this answers is not in the document]
Comment (GabrielDuf) [reply src/UniGetUI.PackageEngine.Managers.WinGet/ClientHelpers/NativeWinGetHelper.cs:355]: Good catch on the second half. The "restored == available" case is handled by the loader's equality filter (same as the CLI paths), but you're right that known-version stuck updates were never suppressed on the COM path, since it doesn't call the suppression logic. Fixed: the COM path now skips a package when IsStuckUpgradeLoop(package) is true, while still avoiding the one-shot "already upgraded" cache (per #5042).

File: src/UniGetUI.PackageEngine.Managers.WinGet/Helpers/WinGetPkgOperationHelper.cs [hunk 4]
```diff
@@ -364,7 +365,11 @@ operation is OperationType.Uninstall
         return OperationVeredict.Failure;
     }
 
-    private static void MarkUpgradeAsDone(IPackage package)
+    // Stop offering an update that "succeeds" this many times without the version ever advancing (#5158).
+    private const int StuckUpgradeThreshold = 3;
+    private const char AttemptSeparator = '\n';
+
+    private static void MarkUpgradeAsDone(IPackage package, bool countTowardStuckLoop = true)
```
[the remark this answers is not in the document]
Comment (GabrielDuf) [reply src/UniGetUI.PackageEngine.Managers.WinGet/Helpers/WinGetPkgOperationHelper.cs:380]: Correct, fixed. MarkUpgradeAsDone now takes countTowardStuckLoop, and the install/update success branches pass operation is OperationType.Update, so installs still refresh the last-installed mark but no longer feed the stuck-update counter.

File: src/UniGetUI.PackageEngine.Managers.WinGet/Helpers/WinGetPkgOperationHelper.cs [hunk 5]
```diff
@@ -379,6 +384,47 @@ private static void MarkUpgradeAsDone(IPackage package)
...
+                && existing[(sep + 1)..] == version
+                && int.TryParse(existing[..sep], out int previous))
+            {
+                count = previous + 1;
+            }
+        }
+        Settings.SetDictionaryItem<string, string>(
+            Settings.K.WinGetUpgradeAttempts,
+            id,
+            $"{count}{AttemptSeparator}{version}"
```
[the remark this answers is not in the document]
Comment (GabrielDuf) [reply src/UniGetUI.PackageEngine.Managers.WinGet/Helpers/WinGetPkgOperationHelper.cs:417]: Valid observation, but this matches the existing WinGetAlreadyUpgradedPackages cache, which is also keyed by Id only — and the one-shot suppression path reads that same cache, so fixing only the new dict wouldn't fully address the identity concern. The ID namespaces across the WinGet sources (winget/msstore/virtual local) don't overlap in practice, so the collision is essentially theoretical. I'd rather not source-qualify one cache and not the other in this PR.

File: src/UniGetUI.PackageEngine.Managers.WinGet/Helpers/WinGetPkgOperationHelper.cs [hunk 6]
```diff
@@ -364,7 +365,11 @@ operation is OperationType.Uninstall
         return OperationVeredict.Failure;
     }
 
-    private static void MarkUpgradeAsDone(IPackage package)
+    // Stop offering an update that "succeeds" this many times without the version ever advancing (#5158).
+    private const int StuckUpgradeThreshold = 3;
```
[the remark this answers is not in the document]
Comment (GabrielDuf) [reply src/UniGetUI.PackageEngine.Managers.WinGet/Helpers/WinGetPkgOperationHelper.cs]: Fixed. Added a WinGetStuckUpgradeThreshold setting (default 3) that StuckUpgradeThreshold now reads, so it matches the description. Added a test covering a non-default value.

File: src/UniGetUI.PackageEngine.Tests/WinGetManagerTests.cs [hunk 7]
```diff
@@ -1115,6 +1115,232 @@ public void ConsumeAlreadyUpgradedSuppression_NoMarkNeverHides()
...
+        Assert.Equal("2.0.0", package.NewVersionString);
+
+        // The mark is not consumed, so it keeps hiding the package on later scans.
+        Assert.Equal(
+            "2.0.0",
+            Settings.GetDictionaryItem<string, string>(
+                Settings.K.WinGetAlreadyUpgradedPackages,
+                "ShiftUp.NIKKE"
+            )
+        );
```
[the remark this answers is not in the document]
Comment (GabrielDuf) [reply src/UniGetUI.PackageEngine.Tests/WinGetManagerTests.cs:1280]: The test verifies the helper's contract — that an already-upgraded unknown package has its installed version restored so VersionString == NewVersionString. The actual hiding is the loader's job (IsPackageValid drops equal-version packages), which is covered separately. So the assertion is intentional, not codifying a phantom; the returned package never reaches the UI as an update.

**Review by mamoreau-devolutions — APPROVED** [review]
